vertical-pod-autoscaler/pkg/recommender/model/vpa.go

```diff
@@ -164,3 +170,45 @@ func (vpa *Vpa) matchesAggregation(aggregationKey AggregateStateKey) bool {
 	}
 	return vpa.PodSelector != nil && vpa.PodSelector.Matches(aggregationKey.Labels())
 }
+
+// UpdateConditions updates the conditions of VPA objects based on it's state.
+// PodsMatched is passed to indicate if there are currently active pods in the
+// cluster matching this VPA.
+func (vpa *Vpa) UpdateConditions(podsMatched bool) {
+	reason := ""
+	msg := ""
+	if !podsMatched {
+		reason = "NoPodsMatched"
+		msg = "No live pods match this VPA object"
+		vpa.Conditions.Set(vpa_types.NoPodsMatched, true, reason, msg)
+	}
+	if vpa.HasRecommendation() {
+		vpa.Conditions.Set(vpa_types.RecommendationProvided, true, "", "")
+	} else {
+		vpa.Conditions.Set(vpa_types.RecommendationProvided, false, reason, msg)
+	}
```

Comment:
Should recommendations continue to be provided on the status if the vpa doesn't match any Pods?

I'm seeing a case where a vpa has no matching pods and still has recommendations, and these recommendations are meaningless (because no pods match to help make the calculations reasonable). So when graphing or viewing data about these vpas we're seeing upperBounds of 30,000mCores as a recommendation for a workload that isn't running at all.

What are your thoughts on adjusting this logic to always remove recommendations after a grace period of time (perhaps 8 days? or the same timeframe as the calculations?) for vpas that have status `NoPodsMatched`?

Reply:
Perhaps here (https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/recommender/model/vpa.go#L145) could be the place to remove ContainerRecommendations for containers that no longer exist or no longer match if the `NoPodsMatch` condition is set?

Let me know your thoughts and I can open a PR or Issue to further discussion :)

Reply:
Ufortunately the NoPodsMatched condition does not work well atm: see #2465, so we can't rely on it.

As good of an idea as it sounds for some class workloads, there are some that would be harmed by removing recommendations that are "old". A good example would be rarely running CronJobs. There is a benefit of having a bootstrap recommendation you can use for new pods even if it is an "old" one.

However, please feel free to open an issue to discuss, perhaps there is a way we can make this work.

Reply:
I was just reading #2465 yesterday as I'm experiencing the same issues in our clusters too. I was reading through the code a bit to try and understand the behaviour of the NoPodsMatched condition, but I haven't gotten far.

As for old recommendations during the NoPodsMatched condition, I don't think CronJobs would benefit from being recommended to have 30000mCores when NoPodsMatched. The behaviour we're seeing is that when NoPodsMatched is set (and there _really_ aren't any Pods) then the recommender is updating the recommendations to be at the extremes with lowerbound <= 10m and upperbound >= 30000m for CPU. 

I agree that having old _correct_ recommendations for rarely running CronJobs would be beneficial as when the CronJob does eventually run the Pod would be admitted with the last good recommendation. However, that's not what I'm seeing at this time. Currently the CronJob's Pod would be admitted with very poor and extreme recommendations.

Reply:
The lowerBound and upperBound are not used to set the actual resource, only target matters. It's a bit unfortunate that they are together in the recommendation object. UpperBound and lowerBound signal the confidence of recommendation and are used only by updater to decide if it should evict running pods. So it does not mean that we recommend 30000 mCPU

Reply:
Oh interesting. I was unable to find much documentation on what the different values mean. We're using https://github.com/FairwindsOps/goldilocks this tool to provide many VPAs for all workloads and then build dashboards off of this information. Goldilocks seems to believe that lowerbound == requests and upperbound == limits, is this not the case?

What should we use from the VPA recommendation as requests and which for limits?

Reply:
Yeah, we need to document this better, it's true.

classical VPA sets request=target and limit in such a way that the original ratio of resource to request is kept constant.

So if you have in a deployment
request=100mCPU
limit=mCPU

VPA will set request=target and limit=2*target

Setting request=lowerBound seems risky, I'd keep request=target. limit=upperBound might work, with the condition that you need to wait until upperBound is reasonably close to target.

Reply:
As a side note, does the Goldilocks tool provide a way to automatically apply the requests or just display the recommended resources?

Reply:
Goldilocks only displays recommendations. We have an internal fork that does more than that and I'm actively working on getting their CLA signed so I can submit patches upstream. So soon™️ 😄